TypeCobol/Compiler/Diagnostics/CrossChecker.cs

```diff
@@ -239,6 +239,83 @@ public override bool Visit(Move move)
             return true;
         }
 
+        public override bool Visit(WhenSearch whenSearch)
+        {
+            var search = whenSearch.Parent;
+            System.Diagnostics.Debug.Assert(search is Search);
+            if (search.CodeElement is SearchBinaryStatement)
+            {
+                var whenSearchVisitor = new WhenSearchVisitor(whenSearch);
+                whenSearch.CodeElement.Condition.AcceptASTVisitor(whenSearchVisitor);
+                if (whenSearchVisitor.IsInError)
+                {
+                    DiagnosticUtils.AddError(whenSearch, "First index declared for the table when subscripting and one of declared keys must be used.");
+                }
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Visit a WhenSearch and set a flag to true if either:
+        /// WHEN does not use one of the declared keys (ascending or descending)
+        /// WHEN does not use the first index declared for the table when subscripting
+        /// </summary>
+        private class WhenSearchVisitor : AbstractAstVisitor
+        {
+            private readonly WhenSearch _whenSearch;
+            public bool IsInError { get; private set; } = false;
+
+
+            public WhenSearchVisitor(WhenSearch whenSearch)
+            {
+                _whenSearch = whenSearch;
+            }
+
+
+            public override bool Visit(NumericVariable numericVariable)
+            {
+                var variableStorageArea = (DataOrConditionStorageArea)numericVariable.StorageArea;
+                // TODO : Handle the case of sub-tables
+                if (variableStorageArea?.Subscripts.Count == 1)
+                {
+                    ////////// WHEN must use one of the declared keys (ascending or descending) //////////
+                    var dataDefinitionKey = _whenSearch.GetDataDefinitionFromStorageAreaDictionary(variableStorageArea);
+                    if (dataDefinitionKey?.Parent.CodeElement is DataDescriptionEntry tableDescriptionEntry)
+                    {
+                        if (tableDescriptionEntry.TableSortingKeys != null)
+                        {
+                            var tableSortingKeyNames = tableDescriptionEntry.TableSortingKeys
+                                .Where(key => key.SortDirection?.Value != SortDirection.None)
+                                .Select(key => key.SortKey.Name);
+                            var isTableSortingKey = tableSortingKeyNames.Contains(dataDefinitionKey.Name);
+                            if (!isTableSortingKey)
+                            {
```

Comment:
Getting the table and its defined keys could be done only once in the visitor constructor ? By examining the search ?

Reply:
It seems to be complicated to examinate the Search and then checking the rule for the WhenSearch as it implies to store elements from the Seach (and handle the case of imbrication...).
This is because we are not in an Analyzer and the children need to be visited for the StoraAreas to be filled.